packages/vike/src/shared-server-client/hooks/execHook.ts

```
@@ -192,10 +204,69 @@ function execHookDirectSync<PageContext extends PageContextPrepareMinimum>(
+  globalContext: GlobalContextPrepareMinimum,
+  pageContext: PageContextPrepareMinimum | null = null,
+): Promise<HookReturn> | HookReturn {
+  return execHookWithOnHookCall(hookFnCaller, hook, globalContext, pageContext, true)
+}
+
+function execHookVanilla<HookReturn>(
+  hookFnCaller: () => HookReturn,
+  hook: Omit<Hook, 'hookTimeout' | 'hookFn'>,
+  globalContext: GlobalContextPrepareMinimum,
+  pageContext: PageContextPrepareMinimum | null = null,
+): Promise<HookReturn> | HookReturn {
+  return execHookWithOnHookCall(hookFnCaller, hook, globalContext, pageContext, false)
+}
+
+function execHookWithOnHookCall<HookReturn>(
+  hookFnCaller: () => HookReturn,
+  hook: Omit<Hook, 'hookTimeout' | 'hookFn'>,
+  globalContext: GlobalContextPrepareMinimum,
+  pageContext: PageContextPrepareMinimum | null,
+  sync = false,
+): HookReturn | Promise<HookReturn> {
+  const { hookName, hookFilePath } = hook
+  assert(hookName !== 'onHookCall') // ensure no infinite loop
+  const configValue = globalContext._pageConfigGlobal.configValues['onHookCall']
+  if (!configValue?.value) return hookFnCaller()
+
+  const context: { pageContext?: unknown; globalContext?: unknown } = { globalContext }
+  if (pageContext) context.pageContext = pageContext
+
+  let originalCalled = false
+  let originalReturn: HookReturn
+  let call: () => HookReturn | Promise<HookReturn> = () => {
+    originalCalled = true
+    originalReturn = hookFnCaller()
+    return originalReturn
+  }
+  for (const onHookCall of configValue.value as Function[]) {
+    const hookPublic = { name: hookName, filePath: hookFilePath, sync, call }
+    // (It would be simpler to define a single hookFnCaller() wrapper instead of chaining call() functions, but it would break the async hooks of the vike-react-sentry extension.)
+    call = () => {
+      // +onHookCall should call hookPublic.call() (the previous call() function) => chaining
+      onHookCall(hookPublic, context)
```

If we don't await this, then when an async hook is wrapped like this:

```js
async onHookCall(hook){
 await 1;
 await hook.call()
}
```

Then `originalReturn` will be undefined

---

> Then `originalReturn` will be undefined

Hm, good catch.

> What do you mean "promise chaining had issues"? Please give an example, I'd like to understand the issue also.

Problems with the version down below:

1. assertUsage() should be called synchronously (but it's called after awaiting the promise instead) — we want to ensure `hook.call()` was called before any `await`.
2. ~~While the `+onHookCall` hooks are *called* sequentially (i.e. they're chained), the promises aren't *awaited* sequentially: Vike (`await execHookVanilla()`) only awaits the promise of the last `+onHookCall` hook.~~ Ah, actually, I'm wrong: the promise of each +onHookCall *are* awaited sequentially. All good here ✅
3.  By design, `+onHookCall` cannot change the return value of a hook. In the same spirit, I'd argue that Vike shouldn't await the promise of `+onHookCall`.

https://github.com/vikejs/vike/blob/a6b912aa1c19b004f34571029fe52b8af54bbd50/packages/vike/src/shared-server-client/hooks/execHook.ts#L261-L268

---

I'm flexible about `3.`, the only problem is `1.`.

So, I think the following version would work? (You can ignore this comment: `ACTUALLY: I'm wrong, the promises aren't awaited sequentially`.)

https://github.com/vikejs/vike/blob/4f9967329b94346e19c9ab0b13800a7b4edf3a55/packages/vike/src/shared-server-client/hooks/execHook.ts#L258-L269

---

> more predictable

Good point. I also like that it's more succinct. Let me revert to that version and fix the CI.

---

As for the CI: 0c3831e98ed35e92232cf8e9082a5d798ad4edc8. WDYT?

---

This is expected for me, swallowing the original hook’s abort,etc.. errors break things.

The question is whether we should allow them to be swallowed.

There are ways to code the wrapping so it can’t swallow the original hook’s errors specifically.

---

The issue is that it returns `Promise<typeof originalReturn>` — Vike internally consumes the wrapped async call, not the original call.

---

How about this: a821d8cb6f8228e55045a8df45e44a011538d695 ?

But note that:
1. It doesn't await the `+onHookCall` promise (unpredictable behavior as you pointed out yesterday)
2. It forces `+onHookCall` to always run `hook.call()` before any `await`

I agree that 1. is a DX regression.

About `2.`, after thinking more about it, I actually now believe it's too brittle to allow `await` before `hook.call()`. Because extensions won't test their extension against `+onBeforeRoute` => when an extension user adds `+onBeforeRoute` to his app => boom :boom:

I'd suggest we address the limitations of `1.` and `2.` only after we get a complaint from a user. If we find a working solution, I'd suggest we just go with it (we can also ship fast and label `+onHookCall` as experimental). But I'm happy to discuss further and feel free to experiment on your own. No rush on my side (although I can't wait to add `+onHookData` to our nice suite of hooks).

---

>[08:23:45.195][/test/playground/test-dev.test.ts][npm run dev][stderr] 8:23:45 AM [vike][request-1][Warning] The onBeforeRoute() hook /+onBeforeRoute.ts returned a promise, but asynchronous routing is deprecated and will be removed in the next major release, see https://vike.dev/route-function#async


I reproduced this warning. 
But the assertion is technically incorrect, because Promise and async function bodies are also executed synchronously.

Explanation:
```ts
const util = require('util')

const p = new Promise((resolve) => resolve(42))
console.log(isFulfilled(p)) // logs true

function isFulfilled(promise) {
  return util.inspect(promise).startsWith('Promise {') && !util.inspect(promise).includes('<pending>')
}
```

```ts
const util = require('util')

const a = async () => {
    return 42
}
console.log(isFulfilled(a())) // logs true

function isFulfilled(promise) {
  return util.inspect(promise).startsWith('Promise {') && !util.inspect(promise).includes('<pending>')
}

```

---

>I'd suggest we address the limitations of 1. and 2. only after we get a complaint from a user.

Okay. Let's go with this.